### ISSUE TITLE
Object list literal in a function

### ISSUE THREAD

**Issue by atifaziz** [gkz/LiveScript#784]
```
[ 1 to 5]
|> map ->
  * x: it
    y: it * 2
  * x: it
    y: it * it
```

Returns:

```
[{"x":1,"y":1},{"x":2,"y":4},{"x":3,"y":9}]
```

Expected:

```
[[{"x":1,"y":2},{"x":1,"y":1}],[{"x":2,"y":4},{"x":2,"y":4}],[{"x":3,"y":6},{"x":3,"y":9}]]
```

Ugly workaround:

```
[ 1 to 3]
|> map ->
    r =
      * x: it
        y: it * 2
      * x: it
        y: it * it
    r
```

Am I misunderstanding something about the syntax for object list literals?

Thanks.

**Comment by summivox** [comment]
I agree that in both cases your code should compile to an (array of dicts) literal.

Minimal breaking example:
```livescript
a = 
  * A: 1
    B: 2
  * C: 3
    D: 4
f = ->
  * A: 1
    B: 2
  * C: 3
    D: 4
```
Seems that lexer returns the same for both. Looking into AST.

**Comment by summivox** [comment]
Starting to believe it's fundamental to how current parser is structured.

Meanwhile, a marginally better fix for your code:
```livescript
[ 1 to 5]
|> map -> return
  * x: it
    y: it * 2
  * x: it
    y: it * it
```

**Comment by vendethiel** [comment]
Yeah, it indeed is how the lexer works... A bit subpar but hey...

**Comment by summivox** [comment]
I realized that function body *is* an implicit list, so implicit return of **last item of the list** is actually the correct semantics... Can't really blame lexer for this :wink:

**Comment by atifaziz** [comment]
@summivox Thanks for the `return` hint. Does read better than having an unused assignment as in the [original workaround I had used](https://github.com/gkz/LiveScript/issues/784#issue-112547975).

**Comment by pepkin88** [comment]
Another way:
```
f = -> []=
  * A: 1
    B: 2
  * C: 3
    D: 4
```

**Comment by vendethiel** [comment]
That's what `*` compiles to

**Comment by pepkin88** [comment]
For the record: `[]=` is just a destructuring assignment without stating a variable, i.e. "anonymous assignment". Do it's an equivalent to the workaround with a temporary variable.

**Comment by johngeorgewright** [comment]
I guess this is the same issue I've just come across:

<table>
  <tr>
    <th>LiveScript</th>
    <th>JavaScript</th>
  </tr>
  <tr>
    <td>
<div class="highlight highlight-source-livescript">
<pre>
table1 =
  * id: 1
    name: 'george'
  * id: 2
    name: 'mike'
  * id: 3
    name: 'donald'

console.log do
  * id: 1
    name: 'george'
  * id: 2
    name: 'mike'
  * id: 3
    name: 'donald'
</pre>
</div>
</td>
<td>
<div class="highlight highlight-source-javascript">
<pre>
// Generated by LiveScript 1.3.1
(function(){
  var table1;
  table1 = [
    {
      id: 1
    }, {
      id: 2
    }, {
      id: 3
    }
  ];
  console.log({
    id: 1
  }, {
    id: 2
  }, {
    id: 3
  });
  console.log({
    id: 1,
    name: 'george'
  }, {
    id: 2,
    name: 'mike'
  }, {
    id: 3,
    name: 'donald'
  });
}).call(this);
</pre>
</div>
</td>
</tr>
</table>

Function calls using the `do` keyword are not correctly converting arrays.

**Comment by rhendric** [comment]
@johngeorgewright Nothing incorrect about that! `do` (in this usage) means ‘treat the following as individual arguments’; `*` means ‘here begins a new structure’. `*` does *not* mean ‘I am an item in an array’ (although lots of people seem to assume that is what it means, and perhaps it should change).

**Comment by johngeorgewright** [comment]
Ah. Apologies, so I need to choose 1 of the following:

```livescript
table1 =
  * id: 1
    name: 'george'
  * id: 2
    name: 'mike'
  * id: 3
    name: 'donald'

console.log table
```

... or ...

```livescript
console.log [
  {
    id: 1
    name: 'mike'
  },
  {
    id: 2
    name: 'mike'
  },
  {
    id: 3
    name: 'donald'
  }
]
```

**Comment by rhendric** [comment]
Or
```ls
console.log [
  * id: 1
    name: 'george'
  * id: 2
    name: 'mike'
  * id: 3
    name: 'donald'
]
```

**Comment by pepkin88** [comment]
Or

```livescript
console.log []=
  * id: 1
    name: 'george'
  * id: 2
    name: 'mike'
  * id: 3
    name: 'donald'
```

**Comment by summivox** [comment]
@rhendric  I believe that this behavior (`func do then (implicit-list)`) should not change. It has become a "feature". We need better documentation that explains the semantics of `*` (hopefully without touching too much compiler internals).
